packages/amplify-graphql-auth-transformer/src/vtl-generator/ddb/resolvers/field.ts

```diff
@@ -204,9 +215,18 @@ export const setDeniedFieldFlag = (operation: string, subscriptionsEnabled: bool
 /**
  * Generates sandbox expression for field
  */
-export const generateSandboxExpressionForField = (sandboxEnabled: boolean): string => {
-  let exp: Expression;
-  if (sandboxEnabled) exp = iff(notEquals(methodCall(ref('util.authType')), str(API_KEY_AUTH_TYPE)), methodCall(ref('util.unauthorized')));
-  else exp = methodCall(ref('util.unauthorized'));
+export const generateSandboxExpressionForField = (sandboxEnabled: boolean, genericIamAccessEnabled: boolean): string => {
+  let exp: Expression = methodCall(ref('util.unauthorized'));
+  if (sandboxEnabled) {
+    exp = iff(notEquals(methodCall(ref('util.authType')), str(API_KEY_AUTH_TYPE)), exp);
+  }
+  if (genericIamAccessEnabled) {
+    const isNonCognitoIAMPrincipal = and([
+      equals(ref('util.authType()'), str(IAM_AUTH_TYPE)),
+      methodCall(ref('util.isNull'), ref('ctx.identity.cognitoIdentityPoolId')),
+      methodCall(ref('util.isNull'), ref('ctx.identity.cognitoIdentityId')),
+    ]);
+    exp = iff(not(parens(isNonCognitoIAMPrincipal)), exp);
+  }
```

Comment:
Can sandbox and IAM access be enabled at the time? It looks like if IAM access is enabled, API Key sandbox expression won't get included in the generated resolver.

Reply:
> Can sandbox and IAM access be enabled at the time?

Yes.

> It looks like if IAM access is enabled, API Key sandbox expression won't get included in the generated resolver.

Sandbox works in such a way that:
1. necessary service directives are added to types and operations
2. `postAuth` resolver is added to handle sandbox auth. See sample `postAuth` resolvers handling sandbox
   1. https://github.com/aws-amplify/amplify-category-api/blob/67290951da610989476d366703b7722229eedbf1/packages/amplify-graphql-auth-transformer/src/__tests__/__snapshots__/iam-access.test.ts.snap#L34-L41
   2. https://github.com/aws-amplify/amplify-category-api/blob/11336ca59d035e5b57b62e11c583158291c9b8b4/packages/amplify-graphql-searchable-transformer/src/__tests__/__snapshots__/generate-resolver-vtl.test.ts.snap#L459-L466
   
   
I can add couple of tests that focus on this particular aspect.

Reply:
added couple of tests here :

https://github.com/aws-amplify/amplify-category-api/pull/2331/commits/355ebb07272c4ecb53f475be76529e0469718fe2

https://github.com/aws-amplify/amplify-category-api/pull/2331/commits/c970c52ed8c9a492b48761ecb00bea37690f327f

Reply:
Thanks for adding more cases. Left a comment to avoid nested auth check.